Issue 1217 - csv upload validation

SageMar opened the pull request:
# Description
"The create meter page stops admins from entering bad values by constraining choices and/or checking before saving. It would be good if the CSV meter upload page did similar checks." After a talk with Steve, it was decided that this validation should be done in JavaScript on the back-end, so these validations are inside the uploadMeters.js file.

Done in collaboration with @cmatthews444 

Partly Addresses #1217

This adds validations for the following variables: 
- Enabled
- Disabled
- Type
- Timezone
- Area
- Cumulative
- Reset
- Timesort
- End Only
- Area Unit
- Min Val
- Max Val
- Disabled Checks

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [ ] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
This does not cover every single validation done on a CSV file - it only covers those listed above.

SageMar committed:
added in the tests for area and meter type. These include a new function for checking each value towards the bottom of the uploadMeters.js file as well as calling them when the submission is sent to the server

SageMar committed:
fixed issue within the added validation for area and meter type, also made sure the correct data was being sent back to the client side

cmatthews444 committed:
Added validation for all boolean value types.

SageMar committed:
added in area unit type checking, still need to adjust to allow for different caps/lowercase

cmatthews444 committed:
Removed old initial testing code.

SageMar committed:
Merge pull request #1 from cmatthews444/Issue1217-csv-validation

Boolean value validations

SageMar committed:
Merge branch 'OpenEnergyDashboard:development' into Issue1217-csv-validation

SageMar committed:
added in a time sort check and fixed some linter errors

cmatthews444 committed:
Add validation for minmax values

cmatthews444 committed:
tweaked timedout

SageMar committed:
Merge pull request #2 from cmatthews444/Issue1217-csv-validation

Min max validation

SageMar committed:
Timezone validation completed

SageMar committed:
added time zone validation

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +293,64 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+		}
+
+		// Validates read values to either false or true
+		if (value !== 'true' && value !== 'false' && value !== true && value !== false) {
```
I know this is a draft PR but a number of the tests were failing. I debugged one of them. The issue is that endOnlyTime on a meter can be undefined. In src/server/sql/meter/create_meters_table.sql it will set it to a default value if undefined. Thus, some tests need to allow for this.

I also noticed that a few inputs go through normalizeBoolean() that allows for yes/no in addition to true/false. FYI.

Please let me know if you need any more information or help.

SageMar replied:
Thank you for those comments, they're very helpful. We took a short break over the holidays so we'll get to work resolving those ASAP. 

Happy new year!

SageMar committed:
clarified error messages and removed trailing spaces. Working on debugging still.

SageMar committed:
Merge branch 'OpenEnergyDashboard:development' into Issue1217-csv-validation

SageMar committed:
Updating to match current OED vers
Merge branch 'Issue1217-csv-validation' of https://github.com/SageMar/OED into Issue1217-csv-validation

SageMar committed:
allow empty min/max values and booleanfields

huss commented:
@SageMar & @cmatthews444 I saw you made some changes since my comments and wanted to see where this PR is at. Are you planning to do more work on this, do you need any help or is that status something else? Thanks.

SageMar commented:
@huss We are definitely still here finishing it up. It's been difficult to find the time between other projects this quarter, but we're aiming to fix the last errors and at least get in the validations we listed in the opening post for this PR. 

When we last left off with the code, we had 3 errors still occurring which we were working to figure out what they mean. It would actually help a ton if you're able to give clarification on them as they're not quite as straightforward as the others we got. If it's easier to chat about this over Discord we can schedule a time on Friday (if you're available, I believe we can both make ourselves available any time on Friday other than 1pm PST), but if they're not too complicated to sum up here that would be great. 

1) Meter testing files starting 'pipe100' doing 'Second meter upload where incorrectly provides meter identifier so fails' with 2 

2) Meter testing files starting 'pipe101' doing 'Second meter with same name so fails but first meter exists' with 1 requests

3) Meter testing files starting 'pipe103' doing 'Uploading meters using Email. First succeeds then the second meter upload fails because it incorrectly provides meter identifier' with 2 requests

huss commented:
@SageMar 

> 1. Meter testing files starting 'pipe100' doing 'Second meter upload where incorrectly provides meter identifier so fails' with 2
> 3. Meter testing files starting 'pipe103' doing 'Uploading meters using Email. First succeeds then the second meter upload fails because it incorrectly provides meter identifier' with 2 requests

Both of these failed because it expected OED to report a success code (200) but it received an error code (400). I would debug this by adding a console.log to the code to print out the response text since that might indicate what went wrong.

> 2. Meter testing files starting 'pipe101' doing 'Second meter with same name so fails but first meter exists' with 1 requests

This one is curious because the returned error is ``Failed to upload meters due to internal OED Error: For meter pipe101 the area entry of 33 is invalid. Area must be a number greater than 0.`` but the value of 33 is greater than 0. I'm guessing you added a test that the area is valid. It may not be working as desired and/or it is returning new info so the test needs to be updated. I would first want to figure out why the test seems to have a message that is inconsistent on 33 > 0.

Do you want to work on these yourselves? I can also help (either working with you or by myself). Let me know what you prefer and if you want/need more information.

SageMar commented:
> Do you want to work on these yourselves? I can also help (either working with you or by myself). Let me know what you prefer and if you want/need more information.

@huss 

I think we'll at least give it a try ourselves with that information. If we're still stuck by Tuesday, I'll reach out again! Thank you!

SageMar committed:
Fixed errors in tests! It was an issue with the way area was being checked, now it converts strings to ints and functions

SageMar commented:
@huss In case you were curious about what caused the issues, it seems the confusing error on pipe101 was what effected all of them. For come reason those files in particular had the number under area coming in as undefined, so using parseInt() on the value ended up fixing all three. 

It passed tests and appears to be working for everything on our end, but let us know if there's anything else that needs to be changed! Thanks for always being so helpful.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -41,21 +42,71 @@ async function uploadMeters(req, res, filepath, conn) {
+			// Verify area unit - have to parse float to ensure it can read in integer numbers like "33"
+			const areaInput = parseFloat(meter[9]);
+			if (areaInput) {
+				if (!isValidArea(areaInput)){
```
This is a small point but here and a few other lines don't have spaces: ``){`` vs ``) }``. Formatting with VSC should clear this up. I can do it if you prefer.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -8,17 +8,18 @@ const Meter = require('../../models/Meter');
 const readCsv = require('../pipeline-in-progress/readCsv');
 const Unit = require('../../models/Unit');
 const { normalizeBoolean } = require('./validateCsvUploadParams');
+const { type } = require('os');
```
I think this include may have been accidentally added as I don't think it is used.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -152,6 +203,78 @@ function switchGPS(gpsString) {
+ * @returns true or false
+ */
+function isValidAreaUnit(areaUnit) {
+	const validTypes = ['feet', 'meters', 'none'];
```
I think using the standard enum would be better. See src/server/models/Unit.js for ``areaUnitType``. This will allow it to adjust if the allowed values are changed. Also, the error msg above with these values should be constructed by looping over the enum. Also for potential future changes the comment(s) listing the values could be updated about the enum rather than specifics.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -152,6 +203,78 @@ function switchGPS(gpsString) {
+ */
+function isValidTimeSort(timeSortValue) {
+	// must be one of the three values
+	if (timeSortValue == 'increasing' || timeSortValue == 'decreasing'){
```
Similar to above about enum, see src/server/services/csvPipeline/validateCsvUploadParams.js for ``MeterTimeSortTypesJS``.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -152,6 +203,78 @@ function switchGPS(gpsString) {
+ * @returns true or false
+ */
+function isValidMeterType(meterTypeString) {
+	const validTypes = ['egauge', 'mamac', 'metasys', 'obvius', 'other'];
```
Same enum and see src/server/models/Meter.js for `Meter.type``.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +293,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
 	return unit.id;
 }
 
-module.exports = uploadMeters;
+
```
Maybe remove second blank line.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -152,6 +203,78 @@ function switchGPS(gpsString) {
+function isValidTimeZone(zone) {
+	// check against the built in timezones, must use a try catch since it does not return a boolean
+	try {
+		new Intl.DateTimeFormat(undefined, {timeZone : zone});
```
The src/client/app/components/TimeZoneSelect.tsx front-end component uses ``moment.tz.names()`` to get the name choices. It would be good to be consistent and use that here if we can. It may mean the try/catch is not needed.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +293,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+ * @param {number} rowIndex - The current row index for error reporting.
+ */
+function validateBooleanFields(meter, rowIndex) {
+// all inputs that involve a true or false all bieng validated together.
```
bieng ->> being

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +293,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+
+	if (isNaN(minValue)){
+		// do nothing, pass it through
+	} else if (isNaN(minValue) || minValue < -9007199254740991 || minValue >= maxValue){
```
I think OED is allowing minValue = maxValue. This applies below.

This did raise the question in my mind that the maxValue is used before it is validated. Should they be checked for isNan, isNumber first, then min/max value and finally if min > max?

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -41,21 +42,71 @@ async function uploadMeters(req, res, filepath, conn) {
+			const timezone = meter[5];
+			if (timezone){
+				if (!isValidTimeZone(timezone)){
+					let msg = `For meter ${meter[0]}, ${timeSortValue} is not a valid time zone.`;
```
I think ``timeSortValue`` should be for time zone instead.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -41,21 +42,71 @@ async function uploadMeters(req, res, filepath, conn) {
 			}
 
+			// Verify area unit - have to parse float to ensure it can read in integer numbers like "33"
+			const areaInput = parseFloat(meter[9]);
```
Just noting you parseFloat before the check for being a number and if it is undefined.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -41,21 +42,71 @@ async function uploadMeters(req, res, filepath, conn) {
 
+			// Verify area unit - have to parse float to ensure it can read in integer numbers like "33"
+			const areaInput = parseFloat(meter[9]);
+			if (areaInput) {
```
This check can fail in certain circumstances. I tested and 0 (zero) and -1 returns false so the area check does not happen. I also notice that 0.1 or any value with a decimal part does not work but it should. Finally, this allowed letters for numbers to pass the test. This may impact other tests you added. I did not try a fix so let me know if you have issues in resolving.

huss reviewed (changes requested):
Thanks to @SageMar & @cmatthews444 for continued work on this issue. I reviewed/tested and made some comments. Please let me know if you need any help with this.

cmatthews444 commented:
@huss We are planning on working on this later this week, will reach and let you know if we need help from there.

SageMar commented:
@huss Do you have any time to meet with us on Friday in the Discord channel? Just looking to get a little clarification on some of the fixes. We're both available anytime after 11AM PST. Thanks!

huss commented:
> @huss Do you have any time to meet with us on Friday in the Discord channel? Just looking to get a little clarification on some of the fixes. We're both available anytime after 11AM PST. Thanks!

@SageMar Sure. Would 12:00 PT work for you?

SageMar commented:
> >
> 
> @SageMar Sure. Would 12:00 PT work for you?

Yup, that works great! See you then

SageMar committed:
fixed spacing + spelling, removed unused import, max value now being checked for validity before calling it to check min<max

SageMar committed:
Enums added where requested. Working on fixing area check now

SageMar committed:
Fixed issues with areaUnit validity check. Zero now returns false and decimals are passing through

huss commented:
@SageMar & @cmatthews444 I wanted to find out the status of this PR. Is it ready or is more work needed? Sorry that I lost track of this.

SageMar committed:
Timezone checks done using moment now

SageMar commented:
> @SageMar & @cmatthews444 I wanted to find out the status of this PR. Is it ready or is more work needed? Sorry that I lost track of this.

Yes, it is ready! I had forgotten to push up the last change to make timezone validation use moment.tz, but that's included now. I believe we fixed all the comments you had before. Thank you for your patience with us!

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +298,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+		// do nothing, pass it through
+	} else if (isNaN(minValue) || minValue < -9007199254740991 || minValue > maxValue) {
+		throw new CSVPipelineError(
+			`Invalid min/max values in row ${rowIndex + 1}: min="${meter[27]}", max="${meter[28]}". ` +
```
Here and below you directly use the meter value. Since you assigned it a variable above I think it would be best to use that to be consistent.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +298,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+	const minValue = Number(meter[27]);
+	const maxValue = Number(meter[28]);
+
+	if (isNaN(minValue) && isNaN(maxValue)) {
```
I feel the logic is tricky. For example, assume only maxValue is a NaN. The first else if will compare minValue > maxValue in the first else if but maxValue is a NaN. Also, the second if below stops the else if so I don't see how the exception can occur in this case. Maybe this logic is better and simpler to understand (open to ideas):

```js
// Check minValue is okay. If maxValue is a NaN then skip last check but it will throw an exception below.
if (isNaN(minValue) || minValue < -9007199254740991 || (!isNaN(maxValue) && minValue > maxValue)) {
    exception
}
// Check maxValue is okay.
// A check for minValue > maxValue is not needed since it would have caused an exception in the
// if above unless minValue or maxValue is a NaN but those cause the exception anyway.
if (isNaN(maxValue) || maxValue > 9007199254740991) {
    exception
}
```

huss replied:
The updated PR has other thoughts on this so it is resolved.

huss commented on src/server/services/csvPipeline/uploadMeters.js:
```diff
@@ -170,4 +298,78 @@ async function getUnitId(unitName, expectedUnitType, conn) {
+		10: 'cumulative',
+		11: 'reset',
+		18: 'end only',
+		32: 'disableChecks'
```
It appears your branch is out of date with development. The definition of disable checks was changed to be an enum. Your code needs to be updated to reflect this. Note you need to merge development and update your DB to do this so let me know if you need help.

huss reviewed (changes requested):
Thanks to @SageMar & @cmatthews444 for continued work on this issue. The changes are good where I made a couple of comments to address.

I did systematic testing of the code and found these checks are not being made:

- gap should be zero or greater.
- variation should be zero or greater.
- duplicates should be 1 to 9.
- if area unit is none then area cannot be specified
- max error must be 0 to 75
- the max date cannot be before the min date; a bonus would be to validate they are valid dates but the check may fail anyway if they are not and that is fine

I realize this is a non-trivial number of changes so let me know if doing them is an issue.

SageMar committed:
Merge branch 'OpenEnergyDashboard:development' into Issue1217-csv-validation

huss commented:
@SageMar & @cmatthews444 I saw your commit on 7 June and appreciate your long efforts on this work. I wanted to see if are or are planning to continue work on this. Please leave a comment with any information. If OED does not hear anything by 1 Aug. then we will try to see about someone else finishing up the needed items to address.

SageMar commented:
> @SageMar & @cmatthews444 I saw your commit on 7 June and appreciate your long efforts on this work. I wanted to see if are or are planning to continue work on this. Please leave a comment with any information. If OED does not hear anything by 1 Aug. then we will try to see about someone else finishing up the needed items to address.

Thank you for your patience with us through this whole process - I had hoped we would be able to do that last round of changes, but unfortunately we have both been extremely busy. It may be best to pass it on to someone else now in order to get it done in a reasonable time frame. 

Thank you again for working with us for so long!

huss commented:
Thanks for the quick response. I've unassigned the issue and will look for someone else to work on this.
